lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -3516,33 +3587,50 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     }
   }
 
-  /// Handle `struct_extract` instruction.
-  ///   y = struct_extract <key>, x
-  ///   adj[x] = struct (0, ..., key: adj[y], ..., 0)
   void visitStructExtractInst(StructExtractInst *sei) {
-    auto *structDecl = sei->getStructDecl();
-    auto av = getAdjointValue(sei);
-    switch (av.getKind()) {
-    case AdjointValue::Kind::Zero:
-      addAdjointValue(sei->getOperand(),
-                      AdjointValue::getZero(sei->getOperand()->getType()));
-      break;
-    case AdjointValue::Kind::Materialized:
-    case AdjointValue::Kind::Aggregate: {
-      SmallVector<AdjointValue, 8> eltVals;
-      for (auto *field : structDecl->getStoredProperties()) {
-        if (field == sei->getField())
-          eltVals.push_back(av);
-        else
-          eltVals.push_back(AdjointValue::getZero(
-              SILType::getPrimitiveObjectType(
-                  field->getType()->getCanonicalType())));
-      }
-      addAdjointValue(sei->getOperand(),
-          AdjointValue::getAggregate(sei->getOperand()->getType(),
-                                     eltVals, allocator));
+    // Replace a `struct_extract` with a call to its pullback.
+    auto loc = remapLocation(sei->getLoc());
+
+    // Get the pullback.
+    auto *pullbackField = getPrimalInfo().lookUpPullbackDecl(sei);
+    if (!pullbackField) {
+      // Inactive `struct_extract` instructions don't need to be cloned into the
+      // adjoint.
+      assert(!activityInfo.isActive(sei, synthesis.indices));
+      return;
     }
+    SILValue pullback = builder.createStructExtract(loc,
+                                                    primalValueAggregateInAdj,
+                                                    pullbackField);
+
+    // Construct the pullback arguments.
+    SmallVector<SILValue, 8> args;
+    auto seed = getAdjointValue(sei);
+    auto *seedBuf = builder.createAllocStack(loc, seed.getType());
+    materializeAdjointIndirectHelper(seed, seedBuf);
```

Comment:
Instead of always allocating stack and emitting a `load` when the adjoint is direct, could you call `materializeAdjointDirect` directly? It'd simplify both your logic and possibly generated code.

Reply:
Oh cool, in fact, I should always be able to `materializeAdjointDirect` because `struct_extract` only works on object types. This simplifies everything.